src/models/bgp/attributes/aspath.rs

```diff
@@ -31,30 +70,318 @@ impl AsPathSegment {
+                        return Some(*asn);
+                    }
+                },
+                AsPathSegment::AsSet(x) => {
+                    self.path = &self.path[1..];
+                    if x.is_empty() {
+                        return Some(Asn::RESERVED);
+                    }
+
+                    let asn = x[(self.route_num % x.len() as u64) as usize];
+                    self.route_num /= x.len() as u64;
+                    return Some(asn);
+                }
+                _ => self.path = &self.path[1..],
+            }
+        }
+    }
+}
+
+pub struct AsPathRouteIter<'a, D> {
+    path: Cow<'a, [AsPathSegment]>,
+    route_num: u64,
+    total_routes: u64,
+    _phantom: PhantomData<D>,
+}
+
+impl<'a, D> Iterator for AsPathRouteIter<'a, D>
+where
+    D: FromIterator<Asn>,
+{
+    type Item = D;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.route_num >= self.total_routes {
+            return None;
+        }
+
+        // Attempt to speed up what is by far the most common case (a path of a single sequence)
+        if self.route_num == 0 && self.path.len() == 1 {
+            if let AsPathSegment::AsSequence(sequence) = &self.path[0] {
+                let route = D::from_iter(sequence.iter().copied());
+                self.route_num += 1;
+                return Some(route);
+            }
+        }
+
+        let route_asn_iter = AsPathNumberedRouteIter {
+            path: self.path.as_ref(),
+            index: 0,
+            route_num: self.route_num,
+        };
+
+        self.route_num += 1;
+        Some(D::from_iter(route_asn_iter))
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Eq, Default, Hash)]
 pub struct AsPath {
     pub segments: Vec<AsPathSegment>,
 }
 
+// Define iterator type aliases. The storage mechanism and by extension the iterator types may
+// change later, but these types should remain consistent.
+pub type SegmentIter<'a> = std::slice::Iter<'a, AsPathSegment>;
+pub type SegmentIterMut<'a> = std::slice::IterMut<'a, AsPathSegment>;
+pub type SegmentIntoIter = std::vec::IntoIter<AsPathSegment>;
+
 impl AsPath {
     pub fn new() -> AsPath {
         AsPath { segments: vec![] }
     }
 
+    /// Shorthand for creating an `AsPath` consisting of a single `AsSequence` segment.
+    pub fn from_sequence<S: AsRef<[u32]>>(seq: S) -> Self {
+        let segment = AsPathSegment::AsSequence(seq.as_ref().iter().copied().map_into().collect());
+
+        AsPath {
+            segments: vec![segment],
+        }
+    }
+
     pub fn from_segments(segments: Vec<AsPathSegment>) -> AsPath {
         AsPath { segments }
     }
 
-    pub fn add_segment(&mut self, segment: AsPathSegment) {
+    // TODO: Would it make more sense to prepend the segment instead of pushing to the end?
+    // Semantically, it feels like it would make more sense to get further from the origin when
+    // pushing a new segment.
+    pub fn push_segment(&mut self, segment: AsPathSegment) {
```

Comment:
Should `push_segment` push the front or back of the `AsPath`? General convention would be to append to the back, but that would change the origin.

Reply:
Why pushing back would change the origin? The origin is expected to be the last ASN of the AS path in this context. Combining with segments, it would be the last segment's last ASN (assuming the last segment is `AsSequence`). I don't want to change this fundamental ordering of the AS path unless necessary.

Thinking from another perspective: the first hop is always the closest to the route collector, and the origin of the prefix is the farthest.

Hope this makes sense.

Reply:
When I say back, I mean the back of the `Vec` of segments in memory, not the back of the path. As it is implemented currently, calling this function will change the origin of the path.

Lets say that I have a path with the segments `[AsSequence([3, 4, 5])]` and I call `push_segment(AsSequence([6]))`. What is the most intuitive outcome?
 - `AsPath { segments: [AsSequence([3, 4, 5]), AsSequence([6])] }` *(`Vec::push(x)` approach)*: This is the current implementation. After this call, the origin ASN changes from 5 to 6.
 - `AsPath { segments: [AsSequence([6]), AsSequence([3, 4, 5])] }` *(`Vec::insert(0, x)` approach)*: This does not change the origin and conceptually matches how a path gets extended as it gets farther from the origin while keeping the segments distinct.
 - `AsPath { segments: [AsSequence([3, 4, 5, 6])] }` *(merge push back approach)*: Same as the `Vec::push(x)` approach, but merges the segments together into a single sequence.
 - `AsPath { segments: [AsSequence([6, 3, 4, 5])] }` *(merge push front approach)*: Same as the `Vec::insert(0, x)` approach, but merges the segments together into a single sequence.

Reply:
This probably falls into the very edge cases. If there were two `AsSequence` (which I doubt it happens in real life), the first scenario should be correct. It probably does not consider as "change of origin" if it is still in the process of parsing all segments. If the `AsSequnce([6])` comes after the other sequence in the MRT, it is closer to the origin. I don't think we should not attempt to merge two `AsSequence` if they appear. I think taking the last `AsSequence` as the AS path is safe.

Reply:
Sounds good. I renamed the function to `append_segment` to hopefully make it a bit more clear and added some separate functions for coalescing and deduping segments of an `AsPath` for if that outcome is desired.